File: concept/17_recursion/binary_search/search/binary_search.dart
import 'dart:io';

void printSearch(List<int> arr, int s, int e) {
  for (var i = s; i < e; i++) {
    stdout.write(arr[i]);
    stdout.write(" ");
  }
}

bool binarySearch(List<int> arr, int s, int e, int key) {
  printSearch(arr, s, e);

  if (s > e) {
    return false;
  }

  int mid = s + (e - s) ~/ 2;
  print("Value of $mid is ${arr[mid]}");

  if (arr[mid] == key) {
    return true;
  }

  if (arr[mid] < key) {
    return binarySearch(arr, mid + 1, e, key);
  } else {
    return binarySearch(arr, s, e - 1, key);
  }
}

void main() {
  int key = 33;
  List<int> arr = [1, 3, 4, 5, 8, 10, 22, 33]; // Sorted

  print(binarySearch(arr, 0, arr.length, key));
}

File: concept/17_recursion/binary_search/search/binary_search.cpp
#include <iostream>
using namespace std;

void print(int arr[], int s, int e)
{
    for (int i = s; i < e; i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl;
}

int binarySearch(int arr[], int s, int e, int key)
{
    print(arr, s, e);
    if (s > e)
    {
        return -1;
    }

    int mid = s + (e - s) / 2;
    cout << "value of mid is : " << arr[mid] << " " << mid << endl;

    if (arr[mid] == key)
    {
        return mid;
    }

    if (arr[mid] < key)
    {
        return binarySearch(arr, mid + 1, e, key);
    }
    else
    {
        return binarySearch(arr, s, e - 1, key);
    }
}

int main()
{
    int key = 22;
    int arr[6] = {1, 3, 4, 5, 8, 22}; // Sorted

    int index = binarySearch(arr, 0, 6, key);
    cout << "At Index : " << index << endl;
}
